Add lexer for Windows Batch files support

carlos-montiers opened the pull request:
Add lexer for support Windows Batch files. This fixes #252.

pyrmont commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+        # Variable Expansions
+        rule %r/[\%!]+([a-z_$@#]+)[\%!]+/i, Name::Variable
+        # For Variables
+        rule %r/(\%+~?[A-Z]+\d?)/i, Keyword::Constant
```
Is this rule correct? Specifically:

1. You didn't mean to create a character class with `\%`?
2. There's no closing delimiter for constants like there is for other variables?
3. You can only have one digit after an alphabetical character?
4. Did you mean for this match to be case-insensitive?

Could you add an example of this to the visual sample?

carlos-montiers replied:
Yes, provided. Is for example: `echo %~dp0`

pyrmont replied:
I felt `Name::Pseudo` was a more appropriate token here and have pushed an updated version to your fork.

pyrmont commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+        # Internal Commands
+        rule %r/@?\b(#{INTERNAL})\s*\b/i, Name::Builtin::Pseudo
+        # External Commands
+        rule %r/\b(#{EXTERNAL})(\.exe|\.com)?\s*\b/i, Name::Builtin::Pseudo
```
Same as above.

carlos-montiers replied:
I changed it and now explicit executables are more generic.

pyrmont commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+        # Lang Keywords
+        rule %r/@?\b(#{KEYWORDS})\s*\b/i, Name::Builtin::Pseudo
+        # Internal Commands
+        rule %r/@?\b(#{INTERNAL})\s*\b/i, Name::Builtin::Pseudo
```
Is there an example of this one in the visual sample?

carlos-montiers replied:
Yes, for example: `@echo off`

pyrmont commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+      end
+
+      state :single_quotes do
+        rule %r/[']/, Str::Single, :pop!
```
There's no way to escape quotes?

carlos-montiers replied:
In the for /f is possible use quotes inside a double quote text:
`for /f %%a in ("-"hi"-") do echo %%~a`
produces:
`-"hi"-`
Is not a common syntax.

pyrmont commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+
+      state :single_quotes do
+        rule %r/[']/, Str::Single, :pop!
+        rule %r/[^']+/, Text
```
Can newlines be included inside single quotes?

carlos-montiers replied:
Very rare. The way is using the caret `^` that indicates the instruction continue in the next line.

pyrmont replied:
As part of the changes I pushed to your fork, I've added an `:escape` state and mixed this state in to the string states to allow for escaped characters (there are examples in the visual sample). Can you have a look and let me know if this looks correct?

carlos-montiers replied:
Yes, it looks correct.

I moved the variable expansions to a mixin. I have doubts about this next: The `%` and `!` characters are used for expand or get the value of the variable, thus I think that characters should be marked as Literal::String::Interpol, but I don't know how implement that.

pyrmont replied:
You can tokenise portions of a match by using capture groups. So if you had the pattern `(a)(b)(c)` you could tokenise it like this:

```ruby
rule %r/(a)(b)(c)/ do |m|
  groups Token1, Token2, Token3
end
```

That said, I'm not sure you really need to do that. I think the way it works now is fine.

pyrmont commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+      end
+
+      state :double_quotes do
+        rule %r/["]/, Str::Double, :pop!
```
Same questions as with `:single_quotes`.

carlos-montiers replied:
Same reply, using the double caret `^^`

pyrmont replied:
Same as with :single_quotes.

pyrmont commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+      end
+
+      state :backtick do
+        rule %r/[`]/, Str::Backtick, :pop!
```
Same questions as with `:single_quotes`.

carlos-montiers replied:
Same, double caret `^^` The backtick is only used as text delimiter (for a command) using the option usebackq in the for /f

pyrmont replied:
Same as with :single_quotes.

pyrmont commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+
+      end
+
+      state :single_quotes do
```
Can you include an example in the visual sample?

carlos-montiers replied:
Included.

pyrmont commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+
+      state :data do
+        rule %r/\s+/, Text
+        rule %r/\^/, Str::Escape
```
Shouldn't this lead into another state? This won't tokenise the characters that presumably come after the caret. (Oh, and example in the visual sample, please.)

carlos-montiers replied:
Yes, the caret indicates that the next character is escaped, and if no character indicates continuation of the instruction in the next line. Too complex for me implement that state. I removed the caret.

pyrmont replied:
I don't know if you want to try mixing in the `:escape` state to the `:data` state in the same way that I did for the string states or if you're happy to leave this out for now.

carlos-montiers replied:
I'm not sure if that is neccesary. Is needed ?

pyrmont replied:
I wasn't sure if escaping is something that is possible 'outside' strings as it were. Usually, I'd say it wouldn't be but in some languages it can (e.g. lisps).

carlos-montiers replied:
Escape a character is possible inside and outside strings.

pyrmont replied:
OK. I've mixed in the `:escape` state to the `:data` state so that it can be used outside strings. I think this is now good to go.

pyrmont commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+        rule %r/[^=\!\%\*\s()$"'`;\\]+/, Text
+      end
+
+      state :paren_inner do
```
There's no route to enter this state.

carlos-montiers replied:
Thanks. I removed it.

pyrmont commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+      state :paren_inner do
+        rule %r/\(/, Operator, :push
+        rule %r/\)/, Operator, :pop!
+        mixin :root
```
If this state mixes in all of `root` and if it has no special rules, is it necessary? (Possibly not since it didn't raise a red flag that the parentheses in the visual sample didn't cause any issues.)

carlos-montiers replied:
I removed that unused state.

carlos-montiers commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+
+      mimetypes 'application/bat', 'application/x-bat', 'application/x-msdos-program'
+
+      KEYWORDS = %w(
```
Done

carlos-montiers commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+
+module Rouge
+  module Lexers
+    class Batch < RegexLexer
```
Done

carlos-montiers commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+
+      state :basic do
+        # Comments
+        rule %r/\b(REM)\b.*$/i, Comment
```
Thanks for the observation. Done.

carlos-montiers commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+        rule %r/:[a-z]+/i, Keyword::Constant
+
+        # Devices
+        rule %r/@?\b(#{DEVICES})\s*\b/i, Keyword::Reserved
```
Thanks for the observation. I removed the whitespace.

carlos-montiers commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+
+        # Devices
+        rule %r/@?\b(#{DEVICES})\s*\b/i, Keyword::Reserved
+        rule %r/@?\b(#{DEVICESNUM}[0-9])\s*\b/i, Keyword::Reserved
```
Yes, there are for old lpt1 and com3 for example.

pyrmont replied:
Right, but it only goes up to 9? There's no LPT10 for example?

carlos-montiers replied:
Yes, only from 0 to 9.
Test in cmd windows 10:
`C:\Users\Carlos>lpt9
' ' is not recognized as an internal or external command,
operable program or batch file.`
These are reserved.
https://www.howtogeek.com/fyi/windows-10-still-wont-let-you-use-these-file-names-reserved-in-1974/

carlos-montiers replied:
Yes, there are: CON, PRN, AUX, NUL, COM1, COM2, COM3, COM4, COM5, COM6, COM7, COM8, COM9, LPT1, LPT2, LPT3, LPT4, LPT5, LPT6, LPT7, LPT8, and LPT9.

https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file

Thus I need fix it from 1-9 not 0-9

carlos-montiers commented on lib/rouge/lexers/batch.rb:
```diff
@@ -0,0 +1,165 @@
+        rule %r/@?\b(#{DEVICESNUM}[0-9])\s*\b/i, Keyword::Reserved
+
+        # Lang Keywords
+        rule %r/@?\b(#{KEYWORDS})\s*\b/i, Name::Builtin::Pseudo
```
I read it and I updated the code with changes.

carlos-montiers commented:
Yes, it looks correct.

I moved the variable expansions to a mixin. I have doubts about this next:
The % and ! characters are used for expand or get the value of the
variable, thus I think that characters should be marked as
Literal::String::Interpol, but I don't know how implement that.

El mar., 30 jul. 2019 a las 16:28, Michael Camilleri (<
notifications@github.com>) escribió:

> *@pyrmont* commented on this pull request.
> ------------------------------
>
> In lib/rouge/lexers/batch.rb
> <https://github.com/rouge-ruby/rouge/pull/1286#discussion_r308927246>:
>
> > +
> +        # Variable Expansions
> +        rule %r/[\%!]+([a-z_$@#]+)[\%!]+/i, Name::Variable
> +        # For Variables
> +        rule %r/(\%+~?[A-Z]+\d?)/i, Keyword::Constant
> +
> +        rule %r/\b(set)\b/i, Keyword::Declaration
> +
> +        # Operators
> +        rule %r/[%!()=<>&|~$*@]/, Operator
> +
> +      end
> +
> +      state :single_quotes do
> +        rule %r/[']/, Str::Single, :pop!
> +        rule %r/[^']+/, Text
>
> As part of the changes I pushed to your fork, I've added an :escape state
> and mixed this state in to the string states to allow for escaped
> characters (there are examples in the visual sample). Can you have a look
> and let me know if this looks correct?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rouge-ruby/rouge/pull/1286?email_source=notifications&email_token=ABLZ2QZSUCYQGQ4V4TR4QADQCCP7RA5CNFSM4IHNKFR2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCABT5BI#discussion_r308927246>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABLZ2Q7XQEPWVBAUDCKTXPLQCCP7RANCNFSM4IHNKFRQ>
> .
>

pyrmont commented on lib/rouge/lexers/batchfile.rb:
```diff
@@ -0,0 +1,163 @@
+        # Labels
+        rule %r/:[a-z]+/i, Name::Label
+
+        rule %r/([a-z]\w*)(\.exe|com|bat|cmd|msi)?/i do |m|
```
@carlos-montiers I'm sorry because I should have brought this up sooner but I realised that, for performance reasons, doing all that regex interpolation isn't the best approach. This approach avoids the interpolation but should produce the same result. The results look fine to me but let me know what you think.

carlos-montiers replied:
Really nice. I like the new version as I learned a new way of do it with more performance. Also I updated the visual sample with code that works correctly, thus is a valid sample.

pyrmont commented:
@carlos-montiers Thanks for the submission of this lexer. Great to close an issue that's been open for more than four years! 🎉
